File: swift/1492_TheKthFactorOfN.swift
class Solution {
    func kthFactor(_ n: Int, _ K: Int) -> Int {
        var divisors = [Int]()
        var sqrtN = Int(sqrt(Double(n))), k = K
        for x in 1...sqrtN {
            if n % x == 0 {
                k -= 1
                divisors.append(x)
                if k == 0 {
                    return x
                }
            }
        }
        
        if sqrtN * sqrtN == n { k += 1 }
        var numDivisors = divisors.count
        return k <= numDivisors ? n / divisors[numDivisors - k] : -1
    }
}

File: python/1492_TheKthFactorOfN.py
class Solution:
    def kthFactor(self, n: int, k: int) -> int:
        divisors = []
        root = floor(sqrt(n)) # sqrt is inflection pt all other divisors are made with ones before it
        for x in range(1, root + 1):
            if n % x == 0:
                k -= 1
                divisors.append(x)
                if k == 0:
                    return x
        
        if root * root == n: k += 1
        return n // divisors[len(divisors) - k] if k <= len(divisors) else -1
        
        
# [1, 2, 3, 4, 6] [9, 12, 18, 36]
